### PR TITLE
Client-side keepalive pings

### PR DESCRIPTION
**Pull request by y-zeng**

### CODE REVIEW

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 1]
```diff
@@ -363,6 +403,29 @@ static void init_transport(grpc_exec_ctx *exec_ctx, grpc_chttp2_transport *t,
...
+                                   INT_MAX});
+        t->keepalive_timeout = value == INT_MAX
+                                   ? gpr_inf_future(GPR_TIMESPAN)
+                                   : gpr_time_from_seconds(value, GPR_TIMESPAN);
+      } else if (0 == strcmp(channel_args->args[i].key,
+                             GRPC_ARG_HTTP2_KEEPALIVE_PERMIT_WITHOUT_CALLS)) {
+        t->keepalive_permit_without_calls =
+            (uint32_t)grpc_channel_arg_get_integer(
+                &channel_args->args[i],
+                (grpc_integer_options){0, 0, MAX_WRITE_BUFFER_SIZE});
```
Comment (sreecha) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: s/MAX_WRITE_BUFFER_SIZE/0

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 2]
```diff
@@ -414,6 +477,16 @@ static void init_transport(grpc_exec_ctx *exec_ctx, grpc_chttp2_transport *t,
   t->ping_state.pings_before_data_required =
       t->ping_policy.max_pings_without_data;
 
+  // Start client-side keepalive pings
```
Comment (sreecha) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Change to C-style comments to be consistent with rest of the file

File: src/core/ext/transport/chttp2/transport/internal.h [hunk 3]
```diff
@@ -382,6 +389,26 @@ struct grpc_chttp2_transport {
   grpc_closure benign_reclaimer_locked;
   /** destructive cleanup closure */
   grpc_closure destructive_reclaimer_locked;
+
+  /* keep-alive ping support */
+  grpc_closure init_keepalive_ping_locked;
+  grpc_closure start_keepalive_ping_locked;
+  grpc_closure finish_keepalive_ping_locked;
+  grpc_closure keepalive_watchdog_fired_locked;
```
Comment (sreecha) [inline src/core/ext/transport/chttp2/transport/internal.h]: Please add comments to these fields.

File: src/core/ext/transport/chttp2/transport/internal.h [hunk 4]
```diff
@@ -382,6 +389,26 @@ struct grpc_chttp2_transport {
...
+  /** time duration in between pings */
+  gpr_timespec keepalive_time;
+  /** grace period for a ping to complete before watchdog kicks in */
+  gpr_timespec keepalive_timeout;
+  /** if keepalive pings are allowed when there's no outstanding streams */
+  bool keepalive_permit_without_calls;
+  /** keep-alive state machine state */
+  grpc_chttp2_keepalive_state keepalive_state;
+
+  uint8_t *keepalive_ping_id;
```
Comment (sreecha) [inline src/core/ext/transport/chttp2/transport/internal.h]: Is `keepalive_ping_id` used anywhere ?

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 5]
```diff
@@ -1987,6 +2076,77 @@ static void finish_bdp_ping_locked(grpc_exec_ctx *exec_ctx, void *tp,
...
+                                            grpc_error *error) {
+  grpc_chttp2_transport *t = arg;
+  if (t->keepalive_state == GRPC_CHTTP2_KEEPALIVE_STATE_PINGING) {
+    if (error == GRPC_ERROR_NONE) {
+      t->keepalive_state = GRPC_CHTTP2_KEEPALIVE_STATE_DYING;
+      close_transport_locked(exec_ctx, t,
+                             GRPC_ERROR_CREATE("keepalive watchdog timeout"));
+    }
+  } else {
+    // The watchdog timer should have been cancelled by
```
Comment (sreecha) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Change to C-style comments (i.e /* */)

File: test/core/end2end/tests/keepalive_timeout.c [hunk 6]
```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ * Copyright 2016, Google Inc.
```
Comment (sreecha) [inline test/core/end2end/tests/keepalive_timeout.c]: s/2016/2017

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c (hunk 1 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Fixed. Thanks a lot for your review!

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c (hunk 2 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Fixed.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c (hunk 5 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: Fixed.

File: src/core/ext/transport/chttp2/transport/internal.h (hunk 3 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/internal.h]: Done.

File: src/core/ext/transport/chttp2/transport/internal.h (hunk 4 above)
Comment (y-zeng) [inline src/core/ext/transport/chttp2/transport/internal.h]: Removed.

File: test/core/end2end/tests/keepalive_timeout.c (hunk 6 above)
Comment (y-zeng) [inline test/core/end2end/tests/keepalive_timeout.c]: Fixed.

**Comment by y-zeng** [comment]
Know issues:
gRPC_interop_pull_requests: #9762, #9797

**Pull request merged**